File: src/main/java/xyz/upperlevel/hgame/world/entity/EntityRegistry.java
package xyz.upperlevel.hgame.world.entity;

import xyz.upperlevel.hgame.input.TriggerInputActionPacket;
import xyz.upperlevel.hgame.network.Endpoint;
import xyz.upperlevel.hgame.world.character.Actor;
import xyz.upperlevel.hgame.world.character.Character;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class EntityRegistry {
    // TODO: use actors instead of entities
    // TODO: fix concurrency issue (both instance an entity at the same time)
    // resolution: only the server can spawn, the client should request the server and then let it spawn
    private List<EntityFactory<?>> factories = new ArrayList<>();
    private Map<Class<?>, Integer> factoryIdByType = new HashMap<>();

    private Map<Integer, Actor> entities = new HashMap<>();
    private int nextId = 0;

    private Endpoint endpoint;

    public Stream<Actor> getEntities() {
        return entities.values().stream();
    }

    private Actor spawn0(int type, float x, float y, boolean left) {
        var factory = factories.get(type);

        var entity = factory.personify(nextId++);
        entity.x = x;
        entity.y = y;
        entity.setLeft(left);
        entities.put(entity.getId(), entity);
        return entity;
    }

    public Actor spawn(Class<? extends Character> entityType, float x, float y, boolean left) {
        int id = factoryIdByType.getOrDefault(entityType, -1);
        if (id == -1) throw new IllegalArgumentException("Entity " + entityType + " not registered!");

        endpoint.send(new EntitySpawnPacket(id, x, y, left));

        return spawn0(id, x, y, left);
    }

    public void registerType(Class<?> type, EntityFactory<? extends Actor> actorFactory) {
        factoryIdByType.put(type, factories.size());
        factories.add(actorFactory);
    }

    public void initEndpoint(Endpoint endpoint) {
        this.endpoint = endpoint;

        endpoint.getEvents().register(EntitySpawnPacket.class, packet -> {
            spawn0(packet.getEntityTypeId(), packet.getX(), packet.getY(), packet.isFacingLeft());
        });
        endpoint.getEvents().register(TriggerInputActionPacket.class, packet -> {
            entities.get(packet.getActorId())
                    .getInput()
                    .onNetworkAction(packet.getActionId());
        });
    }
}

File: src/main/java/xyz/upperlevel/hgame/world/entity/EntityRegistry.kt
package xyz.upperlevel.hgame.world.entity

import xyz.upperlevel.hgame.network.Endpoint
import xyz.upperlevel.hgame.network.NetSide
import xyz.upperlevel.hgame.runSync
import xyz.upperlevel.hgame.world.character.Entity
import xyz.upperlevel.hgame.world.character.Character
import java.util.*
import java.util.stream.Stream

typealias Callback = (Entity) -> Unit

class EntityRegistry {
    // TODO: use actors instead of entities
    private val factories = ArrayList<EntityFactory>()
    private val factoryIdByType = HashMap<Class<*>, Int>()
    private val pendingSpawns = ArrayDeque<Callback>()

    private val _entities = HashMap<Int, Entity>()

    val entities: Stream<Entity>
        get() = _entities.values.stream()

    private var nextId = 0

    private var endpoint: Endpoint? = null

    private fun spawn0(type: Int, x: Float, y: Float, left: Boolean): Entity {
        val factory = factories[type]

        val entity = factory(nextId++)
        entity.body.setTransform(x, y, 0f)
        entity.left = left
        _entities[entity.id] = entity
        return entity
    }

    fun spawn(entityType: Class<out Character>, x: Float, y: Float, left: Boolean, callback: Callback) {
        val id = factoryIdByType[entityType]
            ?: throw IllegalArgumentException("Entity $entityType not registered!")

        endpoint!!.send(EntitySpawnPacket(id, x, y, left, false))

        if (endpoint!!.side === NetSide.MASTER) {
            val actor = spawn0(id, x, y, left)
            callback(actor)
        } else {
            pendingSpawns.add(callback)
        }
    }

    private fun onNetSpawn(typeId: Int, x: Float, y: Float, facingLeft: Boolean, isConfirm: Boolean) {
        val entity = spawn0(typeId, x, y, facingLeft)

        if (endpoint!!.side === NetSide.MASTER) {
            endpoint!!.send(EntitySpawnPacket(typeId, x, y, facingLeft, true))
        } else if (isConfirm) {
            // We are the client and the server sent a confirmation to our request
            pendingSpawns.remove()(entity)
        }
    }

    fun registerType(type: Class<*>, actorFactory: EntityFactory) {
        factoryIdByType[type] = factories.size
        factories.add(actorFactory)
    }

    fun initEndpoint(endpoint: Endpoint) {
        this.endpoint = endpoint

        endpoint.events.register(EntitySpawnPacket::class.java, { packet: EntitySpawnPacket -> runSync { onNetSpawn(packet.entityTypeId, packet.x, packet.y, packet.isFacingLeft, packet.isConfirmation) } })
        endpoint.events.register(TriggerInputActionPacket::class.java, { (actorId, actionId) ->
            runSync {
                _entities[actorId]!!
                        .controller
                        .issue(actionId)
            }
        })
    }
}
